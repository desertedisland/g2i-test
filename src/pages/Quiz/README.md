**Quiz Page**

This page:

- Will retrieve a set of questions / cards from the API
- Cycle through the cards and display the question, in order, one by one  
- Record the user's responses as correct / incorrect in the 'score' store

