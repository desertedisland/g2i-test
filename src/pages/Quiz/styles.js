import styled from 'styled-components';

export const QuizQuestion = styled.p`
  
`;

export const ButtonContainer = styled.div`
  display: flex;
`;
