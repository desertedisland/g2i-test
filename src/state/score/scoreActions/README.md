**Score Actions**

addCorrectAnswer() - Add a correct answer object to the correctAnswers list

addIncorrectAnswer() - Add an incorrect answer object to the incorrectAnswers list

resetAnswers() - Reset the correct and incorrect answer lists to their default state
