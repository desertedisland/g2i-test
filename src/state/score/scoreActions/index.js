const SCORE_ACTIONS = {
  ADD_CORRECT_ANSWER: 'ADD_CORRECT_ANSWER',
  ADD_INCORRECT_ANSWER: 'ADD_INCORRECT_ANSWER',
  RESET_ANSWERS: 'RESET_ANSWERS',
};

// Add a correct answer to the correct answers list
// @param {Object} - answer. An answer or 'card' object
// @returns {Object} - an object to be supplied as an argument to the reducer.
const addCorrectAnswer = (answer) => ({
  type: SCORE_ACTIONS.ADD_CORRECT_ANSWER,
  answer,
});

// Add an incorrect answer to the incorrect answers list
// @param {Object} - answer. An answer or 'card' object
// @returns {Object} - an object to be supplied as an argument to the reducer.
const addIncorrectAnswer = (answer) => ({
  type: SCORE_ACTIONS.ADD_INCORRECT_ANSWER,
  answer,
});

// Reset the correct and incorrect answer lists to their default state
// @returns {Object} - an object to be supplied as an argument to the reducer.
const resetAnswers = () => ({
  type: SCORE_ACTIONS.RESET_ANSWERS,
});

export {
  SCORE_ACTIONS, addCorrectAnswer, addIncorrectAnswer, resetAnswers,
};
