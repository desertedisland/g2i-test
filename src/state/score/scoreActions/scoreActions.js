const SCORE_ACTIONS = {
  ADD_CORRECT_ANSWER: 'ADD_CORRECT_ANSWER',
  ADD_INCORRECT_ANSWER: 'ADD_INCORRECT_ANSWER',
  RESET_ANSWERS: 'RESET_ANSWERS',
};

//
const addCorrectAnswer = (answer) => ({
  type: SCORE_ACTIONS.ADD_CORRECT_ANSWER,
  answer,
});

const addIncorrectAnswer = (answer) => ({
  type: SCORE_ACTIONS.ADD_INCORRECT_ANSWER,
  answer,
});

const resetAnswers = () => ({
  type: SCORE_ACTIONS.RESET_ANSWERS,
});

export {
  SCORE_ACTIONS, addCorrectAnswer, addIncorrectAnswer, resetAnswers,
};
