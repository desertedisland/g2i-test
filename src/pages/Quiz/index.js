import React, { useState, useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { getCardsRequest } from '../../state/cards/cardsActions';
import { addCorrectAnswer, addIncorrectAnswer } from '../../state/score/scoreActions';
import { AlignedPage, HeaderH1, StyledButton } from '../../styles/components';
import { QuizQuestion, ButtonContainer } from './styles';

export default function Quiz() {

  const cards = useSelector((state) => state.cards.cards);
  const working = useSelector((state) => state.cards.working);
  const [cardIndex, setCardIndex] = useState(0);
  let currentCard = cards[cardIndex];
  const dispatch = useDispatch();
  const history = useHistory();

  useEffect(
    () => {

      currentCard = cards[cardIndex];

      if (!cards.length) {

        dispatch(getCardsRequest());

      }

    }, [cards, cardIndex],
  );

  const submitAnswer = (answer) => {

    if (answer.toLowerCase() === currentCard.correct_answer.toLowerCase()) {

      dispatch(addCorrectAnswer(currentCard));

    } else {

      dispatch(addIncorrectAnswer(currentCard));

    }

    if (cardIndex < cards.length - 1) {

      setCardIndex(cardIndex + 1);

    } else {

      // go to the results page if all questions have been answered.
      // eslint-disable-next-line no-unused-expressions
      history && history.push('/results');

    }

  };

  if (working || !currentCard) {

    return <AlignedPage>Please wait...</AlignedPage>;

  }

  return (
    <AlignedPage>

      <HeaderH1 data-test-id="quiz-heading">{currentCard.category}</HeaderH1>

      <QuizQuestion className="quiz-question">
        {currentCard.question}
      </QuizQuestion>

      <ButtonContainer>
        <StyledButton data-test-id="submit-answer-true" onClick={(ev) => submitAnswer('True')}>true</StyledButton>
        <StyledButton data-test-id="submit-answer-false" onClick={(ev) => submitAnswer('False')}>false</StyledButton>
      </ButtonContainer>

    </AlignedPage>
  );

}
