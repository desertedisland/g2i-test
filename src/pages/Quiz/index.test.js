/* eslint no-undef: 0 */

import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { addCorrectAnswer, addIncorrectAnswer } from '../../state/score/scoreActions';
import Quiz from './index';

const mockStore = configureStore([]);

const testStore = {
  cards: {
    working: false,
    cards: [
      {
        category: 'Hitchhikers guide to the galaxy',
        question: 'what is the meaning of life?',
        correct_answer: 'True',
      },
      {
        category: 'Hitchhikers guide to the galaxy',
        question: 'what is the meaning of life?',
        correct_answer: 'True',
      },
    ],
  },
  score: {
    correctAnswers: [],
    incorrectAnswers: [],
  },
};

const store = mockStore(testStore);

describe('Quiz page',
  () => {

    let testRenderer = null;

    beforeAll(
      () => {

        store.dispatch = jest.fn();
        testRenderer = renderer.create(<Provider store={store}><Quiz /></Provider>);

      },
    );

    test('Snapshot test:',
      () => {

        expect(testRenderer.toJSON()).toMatchSnapshot();

      });

    it('Should render the heading correctly',
      () => {

        const heading = testRenderer.root.findByProps({ 'data-test-id': 'quiz-heading' });
        expect(heading.props.children).toEqual('Hitchhikers guide to the galaxy');

      });

    it('Should handle true button click correctly',
      () => {

        renderer.act(
          () => {

            testRenderer.root.findByProps({ 'data-test-id': 'submit-answer-true' }).props.onClick();

          },
        );

        expect(store.dispatch).toHaveBeenCalledTimes(1);
        expect(store.dispatch).toHaveBeenCalledWith(
          addCorrectAnswer(testStore.cards.cards[0]),
        );

      });

    it('Should handle false button click correctly',
      () => {

        renderer.act(
          () => {

            testRenderer.root.findByProps({ 'data-test-id': 'submit-answer-false' }).props.onClick();

          },
        );

        expect(store.dispatch).toHaveBeenCalledTimes(2);
        expect(store.dispatch).toHaveBeenCalledWith(
          addIncorrectAnswer(testStore.cards.cards[1]),
        );

      });

  });
