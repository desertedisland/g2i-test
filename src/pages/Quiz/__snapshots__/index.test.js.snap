// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz page Snapshot test: 1`] = `
<div
  className="sc-gsTCUz fFRaSL"
>
  <h1
    className="sc-bdfBwQ CDlxa"
    data-test-id="quiz-heading"
  >
    Hitchhikers guide to the galaxy
  </h1>
  <p
    className="sc-jSgupP DzdUn quiz-question"
  >
    what is the meaning of life?
  </p>
  <div
    className="sc-gKsewC jmlHni"
  >
    <button
      className="sc-hKgILt ePuoyr"
      data-test-id="submit-answer-true"
      onClick={[Function]}
    >
      true
    </button>
    <button
      className="sc-hKgILt ePuoyr"
      data-test-id="submit-answer-false"
      onClick={[Function]}
    >
      false
    </button>
  </div>
</div>
`;
